File: dotnet/Logify.Alert.Core/Collectors/CompositeInfoCollector.cs
using System;
using System.Collections.Generic;
using System.Runtime.ExceptionServices;

namespace DevExpress.Logify.Core.Internal {
    public abstract class CompositeInfoCollector : IInfoCollector {
        readonly List<IInfoCollector> collectors = new List<IInfoCollector>();

        public IList<IInfoCollector> Collectors { get { return collectors; } }

        protected CompositeInfoCollector(LogifyCollectorContext context) {
            RegisterCollectors(context);
        }

        [HandleProcessCorruptedStateExceptions]
        public virtual void Process(Exception ex, ILogger logger) {
            int count = Collectors.Count;
            for (int i = 0; i < count; i++) {
                try {
                    Collectors[i].Process(ex, logger);
                }
                catch {
                }
            }
        }

        protected abstract void RegisterCollectors(LogifyCollectorContext context);
    }

    public abstract class RootInfoCollector : CompositeInfoCollector {
        protected RootInfoCollector(LogifyCollectorContext context) : base(context) {
        }
    }
}

File: java/Core/src/main/java/com/devexpress/logify/alert/core/collectors/CompositeInfoCollector.java
package com.devexpress.logify.alert.core.collectors;

import com.devexpress.logify.alert.core.logger.ILogger;

import java.util.ArrayList;
import java.util.List;

public abstract class CompositeInfoCollector implements IInfoCollector {
    protected List<IInfoCollector> collectors = new ArrayList<IInfoCollector>();

    protected CompositeInfoCollector() {
    }

    public void process(Throwable ex, ILogger logger) {
        for (IInfoCollector collector : collectors) {
            try {
                collector.process(ex, logger);
            } catch (Exception ignored) {
            }
        }
    }
}
